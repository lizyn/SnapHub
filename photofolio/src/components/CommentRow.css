.comment-row-main {
    display: flex;
    flex-direction: row;
    column-gap: 1rem;
    background-color: white;
    width: 100%;
    padding: 1.3rem 0.4rem 1rem 1rem;
    align-items: center;
    border-top: solid 0.4px lightgray;
    border-bottom: solid 0.4px lightgray;
}

.comment-row-left {
    display: flex;
    flex-direction: row;
    column-gap: 1rem;
    background-color: white;
    width: 80%;
    align-items: center;
}

.comment-row-left p {
    color:rgb(81, 81, 81);
}