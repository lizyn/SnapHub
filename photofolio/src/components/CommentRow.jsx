import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import parse from 'html-react-parser';
import { Avatar } from '@mui/material';
import { fetchUsers } from '../api/axios';
import './CommentRow.css';

function CommentRow(props) {
  CommentRow.propTypes = {
    userId: PropTypes.number.isRequired,
    commentText: PropTypes.string.isRequired
  };

  const { userId, commentText } = props;

  const [commenter, setCommenter] = useState([]);
  const firstRendering = useRef(true);

  useEffect(() => {
    async function fetchCommenterData() {
      const commenters = await fetchUsers(userId);
      setCommenter(commenters);
    }

    if (firstRendering.current) {
      firstRendering.current = false;
      fetchCommenterData();
    }
  });

  return (
    <div className="comment-row-main">
      <Avatar src={commenter.avatar} />
      <p style={{ fontFamily: 'sans-serif' }}>{parse(commentText)}</p>
    </div>
  );
}

export default CommentRow;
