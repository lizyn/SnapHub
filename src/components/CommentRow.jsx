import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import parse from 'html-react-parser';
import IconButton from '@mui/material/IconButton';
import MoreVertIcon from '@mui/icons-material/MoreVert';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import { Avatar } from '@mui/material';
import { Link } from 'react-router-dom';
import { fetchUsers, editComment } from '../api/axios';
import './CommentRow.css';

function CommentRow(props) {
  CommentRow.propTypes = {
    userId: PropTypes.string.isRequired,
    commentText: PropTypes.string.isRequired,
    commentId: PropTypes.string.isRequired,
    commentDel: PropTypes.func.isRequired,
    commentEd: PropTypes.func.isRequired
  };

  const { userId, commentText, commentId, commentDel, commentEd } = props;

  const [anchorEl, setAnchorEl] = useState(false);
  const open = Boolean(anchorEl);
  const [commenter, setCommenter] = useState({});
  const [editing, setEditing] = useState(false);
  const [commentEdit, setCommentEdit] = useState(commentText);

  useEffect(() => {
    setCommenter([]);
    async function fetchCommenterData() {
      const commenters = await fetchUsers(userId);
      setCommenter(commenters[0]);
    }
    fetchCommenterData();
  }, []);

  const handleCommentMenuClick = (e) => {
    setAnchorEl(e.currentTarget);
  };

  const handleEditClose = () => {
    setAnchorEl(null);
  };

  const handleEditClick = () => {
    setEditing(true);
    handleEditClose();
  };

  const handleCommentEdit = (e) => {
    setCommentEdit(e.target.value);
  };

  const handleEditSave = async () => {
    // send edit comment axios request, set await so that refreshing happens after the change
    await editComment(commentId, commentEdit);
    commentEd((currentEdited) => !currentEdited);
    // turn off editing mode
    setEditing(false);
  };

  const handleDeleteClick = () => {
    commentDel(commentId);
    handleEditClose();
  };

  return (
    <div>
      {editing && ( // if editing is true, change comment text area to editing mode
        <div className="comment-row-main">
          <div className="comment-row-left">
            <Link to={`/profile/${userId}`}>
              <Avatar src={commenter.avatar} />
            </Link>
            <input
              type="text"
              placeholder={commentText}
              name="editComment"
              value={commentEdit}
              onChange={handleCommentEdit}
            />
          </div>
          <div>
            <IconButton onClick={handleEditSave}>
              <p>Save</p>
            </IconButton>
          </div>
        </div>
      )}
      {!editing && ( // else just display the comment as usual
        <div className="comment-row-main">
          <div className="comment-row-left">
            <Link to={`/profile/${userId}`}>
              <Avatar src={commenter.avatar} />
            </Link>
            <p style={{ fontFamily: 'sans-serif' }}>{parse(commentText)}</p>
          </div>
          <div>
            <IconButton
              className="comment-row-button"
              id="edit-button"
              aria-controls={open ? 'edit-drop-down' : undefined}
              aria-haspopup="true"
              aria-expanded={open ? 'true' : undefined}
              onClick={handleCommentMenuClick}
            >
              <MoreVertIcon />
            </IconButton>
            <Menu
              id="edit-drop-down"
              anchorEl={anchorEl}
              open={open}
              onClose={handleEditClose}
              MenuListProps={{
                'aria-labelledby': 'edit-button'
              }}
            >
              <MenuItem onClick={handleEditClick}>Edit</MenuItem>
              <MenuItem onClick={handleDeleteClick}>Delete</MenuItem>
            </Menu>
          </div>
        </div>
      )}
    </div>
  );
}

export default CommentRow;
