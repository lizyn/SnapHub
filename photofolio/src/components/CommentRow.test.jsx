/**
 * @jest-environment jsdom
 */
import { React } from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import CommentRow from './CommentRow';

test('Comment row matches snapshot', () => {
  const component = renderer.create(
    <CommentRow
      userId={0}
      commentText=""
      commentId={0}
      commentDel={() => {}}
      commentEd={() => {}}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('click menu button will show edit and delete options', async () => {
  render(
    <CommentRow
      userId={0}
      commentText=""
      commentId={0}
      commentDel={() => {}}
      commentEd={() => {}}
    />
  );
  const editMenu = screen.getByRole('button');
  await userEvent.click(editMenu);
  const editOption = screen.getByRole('menuitem', { name: 'Edit' });
  expect(editOption).toBeInTheDocument();
  const delOption = screen.getByRole('menuitem', { name: 'Delete' });
  expect(delOption).toBeInTheDocument();
});

test('Clicking save when editing will quit editing mode', async () => {
  render(
    <CommentRow
      userId={0}
      commentText=""
      commentId={0}
      commentDel={() => {}}
      commentEd={() => {}}
    />
  );
  const editMenu = screen.getByRole('button');
  await userEvent.click(editMenu);
  const editOption = screen.getByRole('menuitem', { name: 'Edit' });
  await userEvent.click(editOption);
  const saveBtn = screen.getByRole('button', { name: 'Save' });
  expect(saveBtn).toBeVisible();
});

test('Clicking delete will close menu', async () => {
  render(
    <CommentRow
      userId={0}
      commentText=""
      commentId={0}
      commentDel={() => {}}
      commentEd={() => {}}
    />
  );
  const editMenu = screen.getByRole('button');
  await userEvent.click(editMenu);
  const delOption = screen.getByRole('menuitem', { name: 'Delete' });
  await userEvent.click(delOption);
  expect(delOption).not.toBeVisible();
});
