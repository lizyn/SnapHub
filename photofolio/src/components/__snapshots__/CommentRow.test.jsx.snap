// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment row matches snapshot 1`] = `
<div>
  <div
    className="comment-row-main"
  >
    <div
      className="comment-row-left"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
          data-testid="PersonIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
      <p />
    </div>
    <div>
      <button
        aria-haspopup="true"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium comment-row-button css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        id="edit-button"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MoreVertIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
